File: python/timedIO.py
import os
import sys
import time


def read_write_file(file_path, output_path):
    start_read = time.time()
    print("Reading file...")
    with open(file_path, "r") as file:
        lines = file.readlines()
    read_time = (time.time() - start_read) * 1000

    start_write = time.time()
    print("Writing file...")
    with open(output_path, "w") as file:
        for line in lines:
            file.write(line)
    write_time = (time.time() - start_write) * 1000

    print("Done!")
    return read_time, write_time


def main():
    print("\n-----------------------------------")
    print("Python Timed Script")
    print("-----------------------------------\n")

    base_path = sys.argv[1] if len(sys.argv) > 1 else "../"
    input_file = sys.argv[2] if len(sys.argv) > 2 else "biggerSample.txt"
    file_path = os.path.join(base_path, input_file)
    output_file = os.path.join(base_path, "data", "output.txt")

    read_time, write_time = read_write_file(file_path, output_file)

    print(f"Reading time: {read_time:.3f} ms")
    print(f"Writing time: {write_time:.3f} ms")


if __name__ == "__main__":
    main()

File: rust/src/bin/timed_io.rs
use std::env;
use std::fs::File;
use std::io::{self, Read, Write};
use std::path::Path;
use std::time::Instant;

fn read_write_file(input_path: &Path, output_path: &Path) -> io::Result<(u128, u128)> {
    let mut input_file = File::open(input_path)?;
    let mut output_file = File::create(output_path)?;

    let start_read = Instant::now();
    let mut buffer = String::new();
    input_file.read_to_string(&mut buffer)?;
    let read_duration = start_read.elapsed().as_millis();

    let start_write = Instant::now();
    output_file.write_all(buffer.as_bytes())?;
    let write_duration = start_write.elapsed().as_millis();

    Ok((read_duration, write_duration))
}

fn main() -> io::Result<()> {
    println!("\n\n-----------------------------------");
    println!("\nRust Timed Script");
    println!("\n-----------------------------------\n\n");

    let args: Vec<String> = env::args().collect();

    let base_path = if args.len() > 1 { &args[1] } else { "../../../" };

    let input_file = if args.len() > 2 {
        &args[2]
    } else {
        "data/biggerSample.txt"
    };

    let input_file_path = Path::new(base_path).join(input_file);
    let output_file_path = Path::new(base_path).join("data").join("output.txt");

    let (read_time, write_time) = read_write_file(&input_file_path, &output_file_path)?;
    
    println!("Reading time: {} ms", read_time);
    println!("Writing time: {} ms", write_time);

    Ok(())
}
